### PR TITLE
Submission

### PR DESCRIPTION
**Pull request by chelzakaria**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -59,7 +59,29 @@ You are given a set of five images, each of which contains some text in various
...
+!pip install matplotlib
+
+### Installing os
+!pip install os
+
+### Installing keras_ocr
+!pip install keras-ocr
+
+### Installing json
+!pip install json
```
Comment (omarmhaimdat) [inline README.md]: json module is in the standard library

File: README.md [hunk 2]
```diff
@@ -59,7 +59,29 @@ You are given a set of five images, each of which contains some text in various
...
+
+- python3 main.py (for english OCR) 
+- python3 arabicOCR.py (for arabic OCR)
+
+
+### Installing matplotlib
+!pip install matplotlib
+
+### Installing os
+!pip install os
```
Comment (omarmhaimdat) [inline README.md]: os is in the standard library

File: src/arabicOCR.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        
+        # Create the path to the output image
+        out_image = os.path.join(output_folder, file_name +'_bounding_box'+ '.jpg')
+        
+        # Perform OCR on the current image
+        results2 = arabicocr.arabic_ocr(image_path, out_image)
+        
+        # Get the recognized words
+        words=[]
+        for i in range(len(results2)):
```
Comment (omarmhaimdat) [inline src/arabicOCR.py]: `results2` is iterable, which means you can iterate directly on the object like so `for result in results2'

File: output/output.json [hunk 4]
```diff
@@ -0,0 +1,150 @@
+{
+    "input_1": {
```
Comment (omarmhaimdat) [inline output/output.json]: The complete filename was requested as the upper-level node `input_1.png` rather than `input_1`

File: src/arabicOCR.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    # Dictionary to store the OCR results for each image
+    results = {}
+    
+    # Perform OCR on each image in the input folder
+    for image_path in input_images:
+        # Get the file name without the extension
+        file_name = os.path.splitext(os.path.basename(image_path))[0]
+        
+        # Create the path to the output image
+        out_image = os.path.join(output_folder, file_name +'_bounding_box'+ '.jpg')
```
Comment (omarmhaimdat) [inline src/arabicOCR.py]: Why not use f-string for string interpolation like so: `out_image = os.path.join(output_folder, f'{file_name}_bounding_box.jpg')`

File: src/main.py [hunk 6]
```diff
@@ -1 +1,130 @@
...
+        actual_words_set = set(actual_words["words"])
+        predicted_words_set = set(predicted_words_dict[image_name]["words"])
+        
+        # Compare the sets
+        common_words = actual_words_set & predicted_words_set
+        print(f"For image '{image_name}', common words are: {common_words}")
+
+
+#for first section with the original images
+compare_word_sets('./output/output.json')
```
Comment (omarmhaimdat) [inline src/main.py]: This can't be accepted as a valid validation method. The project required that you use a proper validation method such as `WER` and `CER`

**Review by omarmhaimdat — COMMENTED** [review]
Overall this submission looks good. But, I have few comments:

- There is an effort to document the code.
- I am skeptical with the use of `ArabicOcr` since it only uses `easyocr`. I would have preferred to see you use `easyocr` directly.
- Reading files is unnecessarily complicated. You can use `pathlib` to simplify this.
- A lot of code is duplicated between the two scripts. This is not a good practice. You should have a single script that can handle both English and Arabic.
- The code is not very modular. It would be better to have a function that takes an image and returns the text.
- The output is not in the correct format. It should be a JSON file with the following structure:

```json
{
  "input_1.png": [
    "This is a sample text",
    "This is another sample text"
  ], # You can add more lines if needed, or more complex data structures
}
```

- Lack of validation. I have specified in the README that the evaluation will be based on the CER and WER of your solution. I would have liked to see some validation of the results.

In conclusion, really good effort. I would have liked to see a bit more structure and a more modular code. I would have also liked to see some validation of the results.

**Comment by chelzakaria** [comment]
Submission #2

**Comment by omarmhaimdat** [comment]
I will consider your second submission, and you will have my answer on Monday.
